### PR TITLE
Added rows_cols.[pro,alb] to library together with test file rows_col…

### PR DESCRIPTION
**Pull request by kenbowen**
…s_test.pro, doc file rows_cols.md, and ~examples/als/rcx.pro.  Also added test([…]) to arithx1_tests.pro and strings_test.pro.

### CODE REVIEW

**Comment by kenbowen** [comment]
> * Existing I/O builtins put the stream as the first argument. Shouldn't `columns` follow the same pattern? I.e. `columns(Stream, RowList)`

For Input streams that’s ok.  But for output, like columns/2, putting the stream arg first means you can’t take advantage of first argument indexing — which
     output_fixed_rows(ListOfListsOfAtoms, ColsWidths, RowsPads, OutputStream)
would want to do.  And I think in a collection like rows_cols, it’s better to keep everything the same: all last, or all first.  If “all last”, obviously calls on system press will be arg 1st.

> * The u(Headers, LineChar) feature isn't documented, but I wonder if it could be changed. As you know, I'm not a fan of in-band signaling, and it is also very common for the headers to be the first row of a data-table. It would be nice if `columns` could easily handle the first-row headers case. One way might be to use an out-of-band options list. I.e. something like:
> 
> ```
> ?- RowList = [[h1, h2, h3], [1, 2, 3]],
> columns(RowList, [headers(true), widths([5,5,5])]).
> 
> h1    h2   h3
> ==    ==   ==
> 1     2     3
> ```

Somehow lost what (little) doc on it was there.  [Added back]
You need to think about it like markup.   It does more than just headers.
You can create blocks in the table:
city_town_table([
        u(['City', 'Location', 'Population']),
        ['Tokyo', loc(35,139), 9.3],
        ['New York', loc(40,74), 8.6],
        ['Paris', loc(48,24), 2.1],
        u(['Tokyo', loc(23,45), 15.3], 0'_),
        ['Silver City', loc(32,108), 0.000344],
        ['Faro', loc(62,133), 0.01]

cities_towns :- city_town_table(CityData), columns(CityData).

?- cities_towns.
City         Location     Population  
====         ========     ==========  
Tokyo        loc(35,139)  15.3        
New York     loc(40,74)   8.6         
Paris        loc(48,24)   2.1         
_____        __________   ___         
Silver City  loc(32,108)  0.01        
Faro         loc(62,133)  0.000344    

There's no convenient way to do something like that with an options list.
Added this an an example.

> * The tests should cover more of the possible domain of `columns`, including corner cases, and error cases. For example, currently `columns(_)` stack-overflows, and `columns(a)` fails -- they should both throw errors, otherwise they can be very confusing for novice users.

I don’t think the tests (~/core/alsp_src/tests) are for novice users.  A novice user is very unlikely to ever see any of the tests, since they will either install from a tarball, or from an installer (someday).  But ~/core/alsp_src/tests are not included in the tarball.   The tests in ~/core/alsp_src/tests are smoke tests for ALS developers.  I’m willing to guess that in a lot of the other predicates, we don’t test for errors like stack overflow, or all cases of inappropriate input.  Of course, some of our smoke tests are more  thorough than others.
However, novice users ought to see what’s in the examples; hence the BestBooks examples.
Some of your points apply there, e.g., inappropriate input: columns(a) fails, but I don’t think showing columsn(_) hitting stack overflow is important there.  In fact, probably a lot of examples would fail similarly, or behave oddly.

> * If possible, it would be nice to separate the non-column test changes (strings, arithx) from the pull-request. Note that they both need the same fix as for the columns tests (removing cut-printf).

These were consequences of your request that small changes not invoke the full GitHub machinery.  In the case of these two, it seemed sensible to just roll the addition of test([… ) into something that was also doing work in the test directory.

> * I was surprised by the rendering of inline functors. Is this a limitation of `term_codes`? What to do? Example:
> 
> ```
> ?- columns([[line, y = m*x+b], [parabola, y = a*x^2+b*x+c]]).
> line      =(y,+(*(m,x),b))                 
> parabola  =(y,+(+(*(a,^(x,2)),*(b,x)),c))  
> ```

Sigh.  I didn’t think about this.

?- term_codes(23+5, X).
X="+(23,5)"

But fortunately, I can use bufwrite:

?- bufwrite(S, 23+5).
S="23+5"
FIXED

File: examples/als/rcx.pro [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/*======================================================================
+ |                      rows_cols_examples.pro
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply examples/als/rcx.pro]: Nah, I’ll must make the file rows_cols_examples.pro
FIXED

File: core/alsp_src/tests/libtests/rows_cols_test.pro [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	abolish(silly_list/1).
+
+test_var_columns_file :-
+	silly_list(ListOfRows),
+	open('silly.txt', write, S),
+	columns(ListOfRows, S),
+	close(S),
+	grab_lines('silly.txt', SillyColLines),
+	T1 = 'abc    rjtir     x                         y     ',
+	'$strlen'(T1, T1Len),
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/tests/libtests/rows_cols_test.pro]: Just ‘cause I used ’$strlen' everywhere :-).  Old code of mine uses that.  No other reason.  Better to change to atom_length since it’s more modern.   FIXED

File: core/alsp_src/tests/libtests/rows_cols_test.pro [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        	  [abc, rjtir, x, y],
+        	  [yrtbv, h8t, x, yacc],
+        	  [34, f(4,5,6), some_silly(55,yurht, 123), thru]
+        	] )  ),
+	test([
+	test_var_columns_file,
+	test_var_columns_file_uh,
+	test_fxd_columns_file, 
+	test_fxd_columns_file_uh,
+	remove_file('silly.txt'),
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/tests/libtests/rows_cols_test.pro]: FIXED

File: core/alsp_src/tests/libtests/rows_cols_test.pro [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	columns(ListOfRows, S),
+	close(S),
+	grab_lines('silly.txt', SillyColLines),
+	T1 = 'abc    rjtir     x                         y     ',
+	'$strlen'(T1, T1Len),
+	SillyColLines = [L1,L2,L3],
+	'$strlen'(L1, L1Len),
+	T1Len == L1Len,
+	T1 == L1,
+	!.
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/tests/libtests/rows_cols_test.pro]: Good point
FIXED

File: core/alsp_src/library/rows_cols.pro [hunk 5]
```diff
@@ -0,0 +1,567 @@
...
+col_fixed_output(ItemAtom, ItemPad, ColWidth, CurPos, NextPos, S)
+	:-
+	NextPos is CurPos + ColWidth + 2,
+	write(S, ItemAtom),
+	Num is ItemPad + 1,
+	putc_n_of(Num, 0' , S).
+	
+endmod.
+
+/* =================== Samples ===================== */
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: Yes.  Just forgot them.
FIXED

File: core/alsp_src/library/rows_cols.pro [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+	:-
+	(ItemLen > ColWidth ->
+		ItemPad = 0
+		;
+		ItemPad is  ColWidth - ItemLen
+	),
+	pads_for_row(RowItemsLensTail, ColsWidthsTail, RowPadsTail).
+
+
+%put_code(S, 0'|).
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: Hanging there from fooling around with using “|” as an alternate column separator.
Was wondering if I wanted to tackle having an option….
WILL ADD ISSUE TO CONSIDER IT.

File: docs/docs/ref/rows_cols.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    if all of the Rows are of the same length, and if ColWidths  
+    is a list of positive integers of the same length as the  
+    common length as the individual lists in ListOfRows, outputs a  
+    rectangular array of the terms appearing in the Rows to  
+    OutputStream. The array is split into columns of the size of  
+    the corresponding integer of ColWidths, with one character-wide  
+    blank separators.  Terms in Rows which are of size > the  
+    corresponding integer of ColWidths are trucated.  
+
+## EXAMPLES
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply docs/docs/ref/rows_cols.md]: Nice example.  Added a modified version showing what can be done with u(...)

File: docs/docs/ref/rows_cols.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+`columns(ListOfRows)`
+
+`columns(ListOfRows, OutputStream)`
+
+`columns(ListOfRows, ColWidths, OutputStream)`
+
+## DESCRIPTION
+
+**`columns/1`** If ListOfRows is a list of lists of terms (Rows), and
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply docs/docs/ref/rows_cols.md]: That’s because they exactly match the source description lines in rows_cols.pro.
Aesthetically, I’m ok with it, especially ‘cause anything else is going to get into auto or manual line length adjustments/typsetting.  Note that on
        https://cols--alsprolog.netlify.com/docs/ref/rows_cols.html
the description lines are roughly the same length as everything else, from top down to footer.

File: core/alsp_src/library/rows_cols.pro [hunk 9]
```diff
@@ -0,0 +1,567 @@
...
+atoms_for_row([Item | RowTail], CurNumCols, RowNumCols, [ItemAtom | RowAtomsTail], 
+		[ItemLen | RowItemsLen], [MinColWidth | RowMinColWidths])
+	:-
+	atom_for_cols(Item, ItemAtom, ItemLen, MinColWidth),
+	NextNumCols is CurNumCols + 1,
+	atoms_for_row(RowTail, NextNumCols, RowNumCols, RowAtomsTail, RowItemsLen, RowMinColWidths).
+
+atom_for_cols(Item, Item, ItemLen, MinColWidth)
+	:-
+	atom(Item),
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: Probably.  But then why does the system have atom_codes and number_codes?
Though note odd behavior with a UIA:

?- term_codes('Gh#  n89', X).

X="'Gh#  n89'"

yes.
?- term_codes('a', X).

X="a"

File: core/alsp_src/library/rows_cols.pro [hunk 10]
```diff
@@ -0,0 +1,567 @@
...
+
+mk_rows_atoms([Row | RowsTail], NumCols, [RowAtoms | RowsAtomsTail], 
+	 [RowItemsLen | RowsItemLensTail], [RowMinColWidths | RowsMinColsWidthsTail])
+	:-
+	atoms_for_row(Row, 0, RowNumCols, RowAtoms, RowItemsLen, RowMinColWidths),
+	(RowNumCols = NumCols ->
+		mk_rows_atoms(RowsTail, NumCols, RowsAtomsTail, 
+				RowsItemLensTail, RowsMinColsWidthsTail)
+		;
+		throw(rows_len_error(Row,NumCols, RowNumCols))
```
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: Yeah probably.  I was thinking about a last chance handler.
I won't disagree.  However, it's not clear to me what the ErrorTerm should be in:
        throw(error(ErrorTerm, ErrorInfo)).  
The wording of http://127.0.0.1:4000/docs/ref/catch.html (i.e., the list of ErrorTerms) suggests that user code can't throw errors??

File: core/alsp_src/library/rows_cols.pro (hunk 6 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: If there isn't an immediately need for the feature, then you could hold off on creating the issue until the need arrises.

File: core/alsp_src/library/rows_cols.pro (hunk 9 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: `atom_codes` and `number_codes` are part of the ISO standard, but it looks like `term_codes` was proposed but never adopted, so it is an ALS extension.

I gather `term_codes` outputs atoms with quoting on (i.e. as if using `writeq`) -- it doesn't matter if they are interned or un-interned atoms.

File: docs/docs/ref/rows_cols.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply docs/docs/ref/rows_cols.md]: With the latest commit, the problem has gotten worse, with one sentence jutting out (see screenshot below). Similarly, when the text is zoomed on a narrow viewport, the wrapping becomes very ragged (2nd screenshot).

Maintaining these manually line-broken paragraphs will be a hassle when the text is displayed in different layouts, font-sizes, etc. Markdown paragraphs are automatically line-wrapped (when rendered), so they are naturally "layout-agnostic". Wouldn't it be easier to let Markdown wrap most paragraph text for us?

cols--alsprolog netlify com_docs_ref_rows_cols html

image

File: core/alsp_src/library/rows_cols.pro (hunk 10 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: Uneven rows would be a `domain_error` (i.e. the domain of the function  `columns` only includes lists of eqi-length rows, just like `sqrt`'s domain is non-negative floats). You can use the `domain_error/3` helper function to throw the error.

I think it is ok for user code to throw errors, but in this case `columns` is a new builtin, so it should definitely throw appropriate errors like other builtins.

**Comment by kenbowen** [comment]
Couldn’t find this in the current version of https://github.com/AppliedLogicSystems/ALSProlog/pull/161.
So I’m replying to this part alone; 
I’m replying to the rest on https://github.com/AppliedLogicSystems/ALSProlog/pull/161.

See at end:

> On May 29, 2019, at 10:07 AM, Chuck Houpt <notifications@github.com> wrote:
> 
> @chuckhoupt commented on this pull request.
> 
> In core/alsp_src/library/rows_cols.pro:
> 
> > +				RowsItemLensTail, RowsMinColsWidthsTail)
> +		;
> +		throw(rows_len_error(Row,NumCols, RowNumCols))
> +	).
> +
> +atoms_for_row([], RowNumCols, RowNumCols, [], [], []).
> +atoms_for_row([Item | RowTail], CurNumCols, RowNumCols, [ItemAtom | RowAtomsTail], 
> +		[ItemLen | RowItemsLen], [MinColWidth | RowMinColWidths])
> +	:-
> +	atom_for_cols(Item, ItemAtom, ItemLen, MinColWidth),
> +	NextNumCols is CurNumCols + 1,
> +	atoms_for_row(RowTail, NextNumCols, RowNumCols, RowAtomsTail, RowItemsLen, RowMinColWidths).
> +
> +atom_for_cols(Item, Item, ItemLen, MinColWidth)
> +	:-
> +	atom(Item),
> 
> atom_codes and number_codes are part of the ISO standard, but it looks like term_codes was proposed but never adopted, so it is an ALS extension.
> 
> I gather term_codes outputs atoms with quoting on (i.e. as if using writeq) -- it doesn't matter if they are interned or un-interned atoms.

Note that since we’re concerned with lengths, things can get a bit odd if term_codes was used everywhere:

?- term_codes('Fo 5tG', X), length(X, N).

X="'Fo 5tG'" 
N=8 

yes.
?- atom_length('Fo 5tG', N).

N=6 

?- atom_codes('Fo 5tG', X), length(X, N).

X="Fo 5tG" 
N=6

File: core/alsp_src/library/rows_cols.pro (hunk 6 above)
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: Well, I'll make it an enhancement.

File: docs/docs/ref/rows_cols.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (kenbowen) [reply docs/docs/ref/rows_cols.md]: Keeping in mind that:
a.  these lines are mechanically extracted from the source in library/rows_cols.pro;
b.  one does want the source comment in library/rows_cols.pro to look reasonable, as I believe is does now;
c.  part of the goal of the library (and builtins) doc mechanism is to make it easier to end up with documented code (cf. e.g., builtins/filepath.pro and others); it’s better to have some slightly rough-looking doc instead of no doc at all; but certainly as good as is reasonably possible.
Can you make a technical suggestion to maintain a,b&c, and to achieve getting the mechanical extraction to accomplish: "let Markdown wrap most paragraph text” ??

File: core/alsp_src/library/rows_cols.pro (hunk 10 above)
[the remark this answers is not in the document]
Comment (kenbowen) [reply core/alsp_src/library/rows_cols.pro]: I thought something like that at first too.  But the catch/3 page specifies:

domain_error(ValidDomain, Culprit) – The base type of an argument is correct, but the value is outside the domain for which the predicate is defined. The ISO Prolog Standard states that ValidDomain may be any one of the following atoms: character_code_list, character_list, close_option, flag_value, io_mode, not_less_than_zero, operator_priority, operator_specifier, prolog_flag, read_option, source_sink, stream_or_alias, stream_option, stream_position, write_option. Moreover, ALS Prolog allows ValidDomain to take on these additional values : depth_computation, line_length, positive_integer. Culprit is the argument which caused the error.

Unless I’m missing something, none of these domains would cover lists of eqi-length rows.   Where do we state that "sqrt's domain is non-negative floats” ?  I can’t find that, not in REf: is/2.   The domain_error extract above doesn’t list non-negative floats as an error_domain.

I’m not against adding more error domains.  In fact, non-negative floats for use wrt sqrt is probably important, and there are probably others.  And I think that a more “user-code-oriented” error domain ought to be thought about too.  It basically would be available for those situtations we haven’t been able to anticipate (e.g. eqi-length rows).

Also note the following non-error:

?- X is sqrt(-3.4).

X=0nan 

yes.

Maybe we should look into the current state of the ISO standard.  I know the committee still exists.

**Comment by chuckhoupt** [comment]
> Couldn’t find this in the current version of #161.

When Github code comments are resolved (i.e. I clicked the "Resolve Conversation" button), then the code comment appears in collapsed form and one has to click on the "Show resolved" button to expand them.

File: core/alsp_src/library/rows_cols.pro (hunk 10 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: Standard-wize, `ValidDomain` appears to be an open class -- ALS Prolog has already extended it, so the more the merrier.

Sorry about the confusion over `sqrt`. I was making a loose analogy (signaled by "just like"), but I should have said it in a purely mathematical sense:

> ... just like the mathematical function *square root*'s domain is the non-negative reals.
 
(of course, even here the analogy is loose because I don't specify that I'm talking about pre-Bombelli mathematics, which didn't have complex numbers).

The prolog `sqrt` function follows the IEEE floating point standard, which signals domain errors using NaNs.

**Comment by kenbowen** [comment]
Grad school had implicit domains via which course was what: Real function theory vs Analytic function theory.

So, is it time to merge cols and cols-tests into master?
Do I need to go thru the #161 pull request clicking all the outstanding [Resolve conversation] buttons?
—Ken

File: docs/docs/ref/rows_cols.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply docs/docs/ref/rows_cols.md]: Latest commit resolved this, and we can continue discussing doc authoring issues via email.

File: docs/docs/ref/rows_cols.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply docs/docs/ref/rows_cols.md]: Would have preferred to see `city_table` (or something similar) replace `silly_table`, because `silly_table` is "random data", not "toy data", but not a big deal.

**Comment by kenbowen** [comment]
Both city_table and silly_table are there in the EXAMPLES.

**Comment by chuckhoupt** [comment]
> So, is it time to merge cols and cols-tests into master? Do I need to go thru the #161 pull request clicking all the outstanding [Resolve conversation] buttons?

The current open code-comments ($strlen, %putcode, domain_error) are still awaiting code-changes, if you still agree with the suggestions. I'm reviewing the latest changes and will add some comments on those...

File: docs/docs/ref/arithx1.md [hunk 11]
```diff
@@ -38,116 +38,56 @@ predicates:
...
+
+**`min_vector/3`** If Left and Right are both lists of numbers which are the same
+    length and are ground at the time evaluation, then VectorResult is the
+    list of numbers of the same length as Left,Right, and such that
+    the nth element of VectorResult is the min of the nth elements of
+    Left and Right, respectively.
 
 ## EXAMPLES
-```
-?- max(3,92,M1), max(3,50-92,M2), max(-6,-7*9,M3), max(23.89, 23.87,M4).
```
Comment (chuckhoupt) [inline docs/docs/ref/arithx1.md]: The examples that you added in 0c5375c32fb2e9b59cb6f90076b5ca8505baf7ae are removed.

File: docs/docs/ref/html_tokens.md [hunk 12]
```diff
@@ -4,30 +4,28 @@ package: ALS Library
...
+
 `grab_html_tokens(Path, Tokens)`
 
 `read_tokens(S, Tokens)`
 
-`tokenize_file(File, Tokens)`
-
 ## DESCRIPTION
 
-**`grab_html_tokens/2`** Just like `tokenize_file/2`, but wraps read_tokens in unwind_protect
```
Comment (chuckhoupt) [inline docs/docs/ref/html_tokens.md]: The code-quoting changes I made in cc424432488cb83dd388cdc79fdfe2cb84cc8b15 have been lost here.

**Comment by kenbowen** [comment]
> On Jun 1, 2019, at 2:57 PM, Chuck Houpt <notifications@github.com> wrote:
> 
> $strlen, %putcode, domain_error


$strlen:  Changed everywhere to atom_length (FIXED)


%putcode : What’s this about?

domain_error: I’m good with this.

**Comment by kenbowen** [comment]
Do you mean: are removed = should be removed  or are removed = have been removed ??

**Comment by chuckhoupt** [comment]
> Do you mean: are removed = should be removed or are removed = have been removed ??

Sorry, I thought Github would show you the comment in the context of the diff (where a huge chunk of examples are red/deleted). The examples and code quoting used to be in the doc, but where removed by the latest commit. I'm guessing this was accidental. Clearly there's an issue about which doc is the "one true doc", but we should probably hash that out in email.

File: core/alsp_src/tests/libtests/rows_cols_test.pro (hunk 2 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/tests/libtests/rows_cols_test.pro]: Possibly the fix didn't take. I still only see `$strlen`...

File: core/alsp_src/library/rows_cols.pro (hunk 6 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: To resolve this code comment, remove the commented out code:

```
%put_code(S, 0'|).
```

**Comment by kenbowen** [comment]
Confusing.  If I go to branch cols, and go to
	https://github.com/AppliedLogicSystems/ALSProlog/blob/cols/core/alsp_src/library/rows_cols.pro
and search for strlen, I get no hits.

**Comment by kenbowen** [comment]
Just now pushed removing this, on branch cols.

**Comment by kenbowen** [comment]
[Problem located: start at 3rd paragraph]

The only thing I can imagine here is that branches somehow got out of sync.  
html_tokens.md is one of the files modified yesterday in
	Removed double blank md eol’s from descriptions in *.md files, and fo…153cb6f  
I /thought/ my branch cols was up-to-date from master.

Ah, maybe my tool did something wrong.  No. not that.

The problem is that the source for docs/docs/ref/html_tokens.md is alsp_src/library/html_tokens.pro.
The update in cc42443 was applied in docs/docs/ref/html_tokens.md, but not in alsp_src/library/html_tokens.pro.
When I re-ran the tool to get rid of the double-blank-return MD eol’s, html_tokens.pro was one of the files
with double-blank-return MD eol’s, and so the update from cc42443  was lost.

We need to sort this out.  But I suggest for now, that any updates to be applied to a docs/docs/ref/<file>.md, where <file> is a library file, be passed to me to handle.  There’s also a couple of builtins files — like filepath.pro — which are lacking any doc, and I plan to handle them just like the library files.

**Comment by chuckhoupt** [comment]
> Confusing. If I go to branch cols, and go to https://github.com/AppliedLogicSystems/ALSProlog/blob/cols/core/alsp_src/library/rows_cols.pro and search for strlen, I get no hits.

Ah, another issue of context. The original code comment was for the tests (`rows_cols_test.pro`, line 23), which use `$strlen` several times. I hadn't realized you had fixed one instance in the library file, but not in the tests.

File: core/alsp_src/library/rows_cols.pro (hunk 10 above)
[the remark this answers is not in the document]
Comment (chuckhoupt) [reply core/alsp_src/library/rows_cols.pro]: Resolved as: won't fix

**Comment by kenbowen** [comment]
Duh!.  Mindless focus.  Just now pushed the change.

File: docs/docs/ref/arithx1.md (hunk 11 above)
Comment (kenbowen) [reply docs/docs/ref/arithx1.md]: Fixed by  ba8c12a.

File: docs/docs/ref/html_tokens.md (hunk 12 above)
Comment (kenbowen) [reply docs/docs/ref/html_tokens.md]: Fixed by c84f348.
